### DISCUSSION TITLE
How to use MaterialUI's inputRef correctly here?

### DISCUSSION THREAD

**Discussion by leochoo** [react-hook-form/react-hook-form#7229] [Q&A] [answered]
I'm migrating from v6 to v7.
When I click Submit button, it seems that `onSubmit` is not triggered becasue the form is invalid as the `formState` error messages show.
I realized that our whenever our custom component `CustomTextField.tsx` is added to the form, it is causing the issue.

Can someone please take a look at the code above and give me an advice?
I reproduced a minimum viable example here.
https://codesandbox.io/s/testing-inputref-rhf-v7-13hrw?file=/src/App.tsx

 Thank you Bill Luo and all the maintainers for this amazing work.

**Comment by bluebill1049** [comment, marked as answer]
hey @leochoo have a look at this example: https://react-hook-form.com/get-started

Integrating an existing form

```tsx
import React from "react";
import { useForm } from "react-hook-form";

// The following component is an example of your existing Input Component
const Input = ({ label, register, required }) => (
  <>
    <label>{label}</label>
    <input {...register(label, { required })} />
  </>
);

// you can use React.forwardRef to pass the ref too
const Select = React.forwardRef(({ onChange, onBlur, name, label }, ref) => (
  <>
    <label>{label}</label>
    <select name={name} ref={ref} onChange={onChange} onBlur={onBlur}>
      <option value="20">20</option>
      <option value="30">30</option>
    </select>
  </>
));

const App = () => {
  const { register, handleSubmit } = useForm();

  const onSubmit = (data) => {
    alert(JSON.stringify(data));
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <Input label="First Name" register={register} required />
      <Select label="Age" {...register("Age")} />
      <input type="submit" />
    </form>
  );
};
```

you meant to pass down all the props and forward the ref.

or you can use Controller: https://react-hook-form.com/api/usecontroller/controller

**Reply by leochoo** [reply]
@bluebill1049 
I see! Looking at that example,  I simpy added `onChange` field to `CustomTextField` and it started working. I did not have to use controller or forward all ref.

(It seems that `onChange` was not necessary in v6 so i did not realize I had to add it)

Below is the current v7 code.

```typescript
  <CustomTextField
    fullWidth
    label="Title"
    name="title"
    onChange={title.onChange}
    inputRef={title.ref}
    inputProps={{ "aria-label": "title" }}
  />
```

Is this still a valid approach would you say?

**Reply by bluebill1049** [reply]
```
  <CustomTextField
    fullWidth
    label="Title"
    name="title"
    onChange={title.onChange}
    inputRef={title.ref}
    inputProps={{ "aria-label": "title" }}
  />
```

no longer valid for v7.

**Reply by leochoo** [reply]
@bluebill1049  Thank you for the quick responses.

I will try the methods you have suggested using `forward.ref` but I am curious about two things.

1. 
When you said, `no longer valid for v7.`, which part are you referring to?
I never had to use `onChange` in v6 and it worked fine. 
I only added `onChange` after looking at the migration doc.
So my v6 looked like this before.
```
  <CustomTextField
    multiline
    rows={7}
    name="description"
    variant="outlined"
    fullWidth
    inputRef={register({
      required:
        !isAuthorized(authority, userAuthorities.supervisor) ||
        watchPhotos?.length === 0,
    })}
    inputProps={{ "aria-label": "description" }}
    error={!!errors.description}
    disabled={activity.confirmed}
  />
```
2.
My above code with `onChange` works but it is invalid.
I'm wondering how submit can be successful but the form is invalid?
I thought `onSubmit` is not supposed to be triggered when the form is invalid.

 So in my example above [here](https://codesandbox.io/s/testing-inputref-rhf-v7-13hrw?file=/src/App.tsx:1362-1628) .
After I click submit button, console says
```
Pressed 
errors 
{}
isSubmitSuccessful 
true
isValid 
false
```

**Reply by leochoo** [reply]
I wonder if you mistook this as a code i wrote for v6?
This code is actually for v7.
https://github.com/react-hook-form/react-hook-form/discussions/7229#discussioncomment-1762871

**Reply by bluebill1049** [reply]
sorry i mean the register Api at `ref` or `inputRef` is no longer valid

should be something like below

```tsx
  <CustomSelect
    fullWidth
    label="Title"
    name="title"
    {...register('inputname')}
  />

const CustomSelect = React.forwardRef(({ onChange, onBlur, name, label }, ref) => (
  <>
    <label>{label}</label>
    <select name={name} ref={ref} onChange={onChange} onBlur={onBlur}>
      <option value="20">20</option>
      <option value="30">30</option>
    </select>
  </>
));
  ```

**Reply by leochoo** [reply]
Thank you Bill! Got it!

**Reply by bluebill1049** [reply]
np 👍
